WIP: Restart dunst on configuration changes

flokli opened the pull request:
This changes the dunst service to restart on config change.

Currently, systemd doesn't really notice socket-activated dunst processes and would spawn a second one on restart, so we `pkill` the old one for now.

nonsequitur commented:
Maybe I'm missing something obvious, but I'm using [this simple patch](https://github.com/nonsequitur/home-manager/commit/92fd219e5c70360f80a5b035d0084e37db73b6aa) together with my [auto start services PR](https://github.com/rycee/home-manager/pull/164/) without any issues of extra processes being spawned.

rycee commented on modules/services/dunst.nix:
```diff
@@ -85,75 +89,87 @@ in
-                ) basePaths
-              ) themes
+    {
+      xdg.configFile."dunst/dunstrc".source = configFile;
```
This will unconditionally write the `dunstrc` file, which we don't want if the user does not set any settings since they then probably have they own `dunstrc`.

rycee commented on modules/services/dunst.nix:
```diff
@@ -85,75 +89,87 @@ in
         };
+      };
+      home.activation.checkDunst = dagEntryBefore [ "linkGeneration" ] ''
+        if ! cmp --quiet \
```
This should only happen if the user actually manage the configuration file through this module.

flokli commented:
@rycee sorry for letting this rot around for so long! I will adress the comments from above shortly.

Do we still need to `pkill`, or was there some change in [the NixOS session dbus](https://github.com/rycee/home-manager/issues/122#issuecomment-342987129)?

flokli commented:
while rebasing, noticed the changes to use xdg module were already included in d70715a635461e77d4426ed50e363a12066fe0a9 - but restarting currently still isn't working.

flokli commented:
@rycee I rebased, moved the `pkill` call to the `ExecStop` systemd dunst unit option, and added dunstrc to `X-Restart-Triggers` in case home-manager manages dunst configuration. This now causes home-manager to suggest restaring `dunst` if `systemd.user.startServices = false;` when its configuration is updated:

```
Jan 12 16:36:24 tp hm-activate-flokli[32608]: Starting home manager activation
Jan 12 16:36:24 tp hm-activate-flokli[32608]: Activating checkFilesChanged
Jan 12 16:36:24 tp hm-activate-flokli[32608]: Activating checkLinkTargets
Jan 12 16:36:24 tp PackageKit[32609]: daemon quit
Jan 12 16:36:24 tp systemd[1]: Started PackageKit Daemon.
Jan 12 16:36:24 tp hm-activate-flokli[32608]: Activating writeBoundary
Jan 12 16:36:24 tp hm-activate-flokli[32608]: Activating installPackages
Jan 12 16:36:24 tp hm-activate-flokli[32608]: replacing old 'home-manager-path'
Jan 12 16:36:24 tp hm-activate-flokli[32608]: installing 'home-manager-path'
Jan 12 16:36:24 tp nix-daemon[1116]: accepted connection from pid 32671, user flokli
Jan 12 16:36:24 tp hm-activate-flokli[32608]: Activating linkGeneration
Jan 12 16:36:24 tp hm-activate-flokli[32608]: Cleaning up orphan links from /home/flokli
Jan 12 16:36:24 tp hm-activate-flokli[32608]: Creating profile generation 153
Jan 12 16:36:24 tp hm-activate-flokli[32608]: Creating home file links in /home/flokli
Jan 12 16:36:24 tp hm-activate-flokli[32608]: Activating onFilesChange
Jan 12 16:36:24 tp hm-activate-flokli[32608]: Activating reloadSystemD
Jan 12 16:36:24 tp dbus-daemon[22617]: [session uid=1000 pid=22617] Reloaded configuration
Jan 12 16:36:24 tp dbus-daemon[2428]: [session uid=1000 pid=2420] Reloaded configuration
Jan 12 16:36:24 tp systemd[2402]: Reloading.
Jan 12 16:36:25 tp hm-activate-flokli[32608]: Suggested commands:
Jan 12 16:36:25 tp hm-activate-flokli[32608]: systemctl --user restart dunst.service
Jan 12 16:36:25 tp hm-activate-flokli[32608]: Activating xdgCreateCache
Jan 12 16:36:25 tp systemd[1]: Started Home Manager environment for flokli.
```

This doesn't restart automatically, even though that's safe. Should I add an activation script doing that, like in the previous PR iteration?

From my understanding, setting `systemd.user.startServices = true;`, triggering the new-style `systemd-activate.rb` script should already restart it, but it doesn't seem to pick up the change at all:

```
Jan 12 16:45:41 tp hm-activate-flokli[14536]: Activating checkLinkTargets
Jan 12 16:45:41 tp hm-activate-flokli[14536]: Activating writeBoundary
Jan 12 16:45:41 tp hm-activate-flokli[14536]: Activating installPackages
Jan 12 16:45:41 tp hm-activate-flokli[14536]: replacing old 'home-manager-path'
Jan 12 16:45:41 tp hm-activate-flokli[14536]: installing 'home-manager-path'
Jan 12 16:45:41 tp nix-daemon[1116]: accepted connection from pid 14599, user flokli
Jan 12 16:45:41 tp hm-activate-flokli[14536]: Activating linkGeneration
Jan 12 16:45:41 tp hm-activate-flokli[14536]: Cleaning up orphan links from /home/flokli
Jan 12 16:45:41 tp hm-activate-flokli[14536]: Creating profile generation 156
Jan 12 16:45:41 tp hm-activate-flokli[14536]: Creating home file links in /home/flokli
Jan 12 16:45:41 tp hm-activate-flokli[14536]: Activating onFilesChange
Jan 12 16:45:41 tp hm-activate-flokli[14536]: Activating reloadSystemD
Jan 12 16:45:41 tp dbus-daemon[2428]: [session uid=1000 pid=2420] Reloaded configuration
Jan 12 16:45:41 tp dbus-daemon[22617]: [session uid=1000 pid=22617] Reloaded configuration
Jan 12 16:45:41 tp systemd[2402]: Reloading.
Jan 12 16:45:41 tp hm-activate-flokli[14536]: Starting: setxkbmap.service
Jan 12 16:45:41 tp systemd[2402]: Starting Set up keyboard in X...
Jan 12 16:45:41 tp systemd[2402]: Started Set up keyboard in X.
Jan 12 16:45:41 tp hm-activate-flokli[14536]: Activating xdgCreateCache
Jan 12 16:45:41 tp systemd[1]: Started Home Manager environment for flokli.
```

rycee commented:
The systemd service here is mostly for non-NixOS distros that have a more functional user dbus session setup. And I'm not certain it is functioning well there either. Might be best to completely remove it as it is pretty confusing to have around.

My idea would be to apply this change instead:
```patch
diff --git a/modules/services/dunst.nix b/modules/services/dunst.nix
index 0e82d43..0a5daf3 100644
--- a/modules/services/dunst.nix
+++ b/modules/services/dunst.nix
@@ -155,7 +155,17 @@ in
       }
 
       (mkIf (cfg.settings != {}) {
-        xdg.configFile."dunst/dunstrc".text = toDunstIni cfg.settings;
+        xdg.configFile."dunst/dunstrc" = {
+          text = toDunstIni cfg.settings;
+          onChange = ''
+            pkillVerbose=""
+            if [[ -v VERBOSE ]]; then
+              pkillVerbose="-e"
+            fi
+            $DRY_RUN_CMD ${pkgs.procps}/bin/pkill -u $USER $pkillVerbose dunst
+            unset pkillVerbose
+          '';
+        };
       })
     ]
   );
```

rycee commented:
I created #533 with my proposed change. I'm uncertain whether it is save to fix the service definition to include the `X-Restart-Triggers` since it might cause people to attempt using the service on NixOS 😕

flokli commented:
#533 fixed restarting dunst on configuration changes for me. Thanks!
